File: mobile/app/src/main/java/com/skladpro/demo/MainActivity.java
package com.skladpro.demo;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.Uri;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.webkit.ValueCallback;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.Button;
import android.widget.ProgressBar;

/**
 * Демо-обёртка SkladPro: WebView открывает сайт.
 * Обрабатывает Back, ошибки сети/offline, загрузку файлов (аватар),
 * и переживает поворот экрана (configChanges в манифесте).
 */
public class MainActivity extends Activity {

    private static final int FILE_CHOOSER_CODE = 1001;

    private WebView webView;
    private View offlineView;
    private ProgressBar progress;
    private ValueCallback<Uri[]> fileCallback;
    private boolean pageError = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        // Переключаемся со splash-темы на основную сразу после старта.
        setTheme(R.style.AppTheme);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        webView = findViewById(R.id.webview);
        offlineView = findViewById(R.id.offline_view);
        progress = findViewById(R.id.progress);
        Button retry = findViewById(R.id.retry_button);
        retry.setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View v) { loadStart(); }
        });

        WebSettings s = webView.getSettings();
        s.setJavaScriptEnabled(true);
        s.setDomStorageEnabled(true);          // localStorage: сайт хранит JWT здесь
        s.setDatabaseEnabled(true);
        s.setMediaPlaybackRequiresUserGesture(false);
        s.setAllowFileAccess(true);
        s.setLoadWithOverviewMode(true);
        s.setUseWideViewPort(true);
        s.setSupportZoom(false);
        s.setCacheMode(WebSettings.LOAD_DEFAULT);

        webView.setWebViewClient(new WebViewClient() {
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
                return false; // навигация остаётся внутри WebView
            }
            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {
                pageError = false;
                progress.setVisibility(View.VISIBLE);
            }
            @Override
            public void onPageFinished(WebView view, String url) {
                progress.setVisibility(View.GONE);
                if (!pageError) {
                    offlineView.setVisibility(View.GONE);
                    webView.setVisibility(View.VISIBLE);
                }
            }
            @Override
            public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
                if (request.isForMainFrame()) {
                    pageError = true;
                    showOffline();
                }
            }
        });

        webView.setWebChromeClient(new WebChromeClient() {
            @Override
            public void onProgressChanged(WebView view, int newProgress) {
                progress.setProgress(newProgress);
            }
            @Override
            public boolean onShowFileChooser(WebView view, ValueCallback<Uri[]> callback,
                                             FileChooserParams params) {
                if (fileCallback != null) fileCallback.onReceiveValue(null);
                fileCallback = callback;
                try {
                    startActivityForResult(params.createIntent(), FILE_CHOOSER_CODE);
                } catch (Exception e) {
                    fileCallback = null;
                    return false;
                }
                return true;
            }
        });

        if (savedInstanceState == null) {
            loadStart();
        } else {
            webView.restoreState(savedInstanceState);
        }
    }

    private void loadStart() {
        if (isOnline()) {
            offlineView.setVisibility(View.GONE);
            webView.setVisibility(View.VISIBLE);
            webView.loadUrl(getString(R.string.app_url));
        } else {
            showOffline();
        }
    }

    private void showOffline() {
        webView.setVisibility(View.GONE);
        progress.setVisibility(View.GONE);
        offlineView.setVisibility(View.VISIBLE);
    }

    private boolean isOnline() {
        ConnectivityManager cm = (ConnectivityManager) getSystemService(CONNECTIVITY_SERVICE);
        if (cm == null) return false;
        NetworkInfo ni = cm.getActiveNetworkInfo();
        return ni != null && ni.isConnected();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == FILE_CHOOSER_CODE) {
            if (fileCallback != null) {
                fileCallback.onReceiveValue(
                        WebChromeClient.FileChooserParams.parseResult(resultCode, data));
                fileCallback = null;
            }
            return;
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK && webView != null && webView.canGoBack()) {
            webView.goBack();   // аппаратный Back = назад по истории WebView
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        webView.saveState(outState);
    }
}

File: mobile/android-demo/app/src/main/java/com/skladpro/demo/MainActivity.kt
package com.skladpro.demo

import android.annotation.SuppressLint
import android.net.ConnectivityManager
import android.net.Uri
import android.net.http.SslError
import android.os.Bundle
import android.webkit.SslErrorHandler
import android.webkit.ValueCallback
import android.webkit.WebChromeClient
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.activity.OnBackPressedCallback
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import com.skladpro.demo.databinding.ActivityMainBinding

/**
 * Демо-обёртка WebView над сайтом SkladPro.
 *
 * Отвечает за требования ТЗ этапа L:
 *  - открывает сайт (URL из res/values/strings.xml -> site_url);
 *  - обрабатывает системную кнопку Back (навигация внутри WebView);
 *  - обрабатывает offline и сетевые ошибки (экран с кнопкой «Повторить»);
 *  - поддерживает загрузку файлов (аватар/фото в приложении);
 *  - pull-to-refresh.
 *
 * JavaScript и DOM storage включены НАМЕРЕННО: SPA хранит JWT в localStorage,
 * без domStorage вход не работал бы.
 */
class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private var filePathCallback: ValueCallback<Array<Uri>>? = null
    private var lastFailed = false

    // Диалог выбора файла для <input type="file"> (загрузка фото/аватара).
    private val fileChooser =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            val callback = filePathCallback ?: return@registerForActivityResult
            callback.onReceiveValue(
                WebChromeClient.FileChooserParams.parseResult(result.resultCode, result.data)
            )
            filePathCallback = null
        }

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreate(savedInstanceState: Bundle?) {
        // Сбрасываем splash-тему на основную до отрисовки контента.
        setTheme(R.style.Theme_SkladPro)
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupWebView()
        setupBackHandling()

        binding.swipeRefresh.setOnRefreshListener { reload() }
        binding.retryButton.setOnClickListener { reload() }

        if (savedInstanceState == null) loadStart()
        else binding.webView.restoreState(savedInstanceState)
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun setupWebView() = with(binding.webView.settings) {
        javaScriptEnabled = true
        domStorageEnabled = true          // localStorage для JWT
        databaseEnabled = true
        loadWithOverviewMode = true
        useWideViewPort = true
        cacheMode = android.webkit.WebSettings.LOAD_DEFAULT
        allowFileAccess = true

        binding.webView.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(
                view: WebView, request: WebResourceRequest
            ): Boolean = false  // всё грузим внутри WebView

            override fun onPageStarted(view: WebView?, url: String?, favicon: android.graphics.Bitmap?) {
                lastFailed = false
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                binding.swipeRefresh.isRefreshing = false
                if (!lastFailed) showContent()
            }

            override fun onReceivedError(
                view: WebView, request: WebResourceRequest, error: WebResourceError
            ) {
                // Только для основного документа, а не для под-ресурсов.
                if (request.isForMainFrame) {
                    lastFailed = true
                    showError()
                }
            }

            override fun onReceivedSslError(
                view: WebView?, handler: SslErrorHandler, error: SslError?
            ) {
                // Демо: НЕ игнорируем SSL-ошибки (безопасность). Отменяем загрузку.
                handler.cancel()
                lastFailed = true
                showError()
            }
        }

        binding.webView.webChromeClient = object : WebChromeClient() {
            override fun onShowFileChooser(
                webView: WebView?,
                callback: ValueCallback<Array<Uri>>?,
                params: FileChooserParams?
            ): Boolean {
                filePathCallback?.onReceiveValue(null)
                filePathCallback = callback
                val intent = params?.createIntent()
                if (intent == null) {
                    filePathCallback = null
                    return false
                }
                return try {
                    fileChooser.launch(intent)
                    true
                } catch (e: Exception) {
                    filePathCallback = null
                    false
                }
            }
        }
    }

    private fun setupBackHandling() {
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                if (binding.webView.canGoBack()) binding.webView.goBack()
                else finish()
            }
        })
    }

    private fun loadStart() {
        if (isOnline()) {
            binding.webView.loadUrl(getString(R.string.site_url))
        } else {
            lastFailed = true
            showError()
        }
    }

    private fun reload() {
        if (isOnline()) {
            showContent()
            binding.webView.reload()
        } else {
            binding.swipeRefresh.isRefreshing = false
            showError()
        }
    }

    private fun isOnline(): Boolean {
        val cm = getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
        val network = cm.activeNetwork ?: return false
        val caps = cm.getNetworkCapabilities(network) ?: return false
        return caps.hasCapability(android.net.NetworkCapabilities.NET_CAPABILITY_INTERNET)
    }

    private fun showError() {
        binding.swipeRefresh.isRefreshing = false
        binding.errorView.visibility = android.view.View.VISIBLE
        binding.swipeRefresh.visibility = android.view.View.GONE
    }

    private fun showContent() {
        binding.errorView.visibility = android.view.View.GONE
        binding.swipeRefresh.visibility = android.view.View.VISIBLE
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        binding.webView.saveState(outState)
    }
}
